packages/dds/tree/docs/user-facing/schema-evolution.md

```diff
@@ -0,0 +1,304 @@
+information about that schema is persisted as part of the document.
+The schema that the application works with in code is hereforth referred to as "view schema" and the schema persisted with the document "stored schema."
+
+Compatibility scenarios can be described in terms of discrepancies between the view schema and the stored schema.
+For example, an application in the geometry domain might try to open a document using a view schema constructed with:
+
+```typescript
+const builder = new SchemaFactory("Geometry");
+
+class Point extends builder.object({
+	x: builder.number,
+	y: builder.number,
+}) {}
+
+class Circle extends builder.object({
+	center: Point,
+	radius: builder.number,
+}) {}
+
+class Plane extends builder.object({
+	shapes: builder.array([Circle, Point]),
+}) {}
+```
+
+but discover that the document's stored schema doesn't allow `Point` children under the `shapes` field, i.e. its definition of `Plane` was this:
+
+```typescript
+class Plane extends builder.object({
+	shapes: builder.array([Circle]),
+}) {}
+```
+
+Applications may leverage `SharedTree`'s compatibility APIs to implement policies that allow clients running incompatible code to fail in predictable ways.
+
+## General Approach
+
+SharedTree allows application authors to make _backward-compatible changes to document schema_ using `TreeView.upgradeSchema`.
+A change to document schema is backward-compatible if the set of documents allowed by the old schema is a subset of documents allowed by the new schema.
+Put differently, valid documents must remain valid.
+
+The following types of changes are backward-compatible under this definition:
+
+1. Adding an optional field to an object node
+2. Adding to the set of allowed types for a field
+3. Relaxing a field kind to a more general field kind
+
+Adding new schemas (rather than modifying existing ones) is allowed and a typical reason to make a change to the allowed types for a field.
+Examples for each type of change can be found below.
+Each change should be interpreted independently from the others (rather than in aggregate):
+
+```typescript
+// Base schema:
+const builder = new SchemaFactory("Geometry");
+
+class Point extends builder.object({
+	x: builder.number,
+	y: builder.number,
+}) { }
+
+class Circle extends builder.object({
+	center: Point,
+	radius: builder.number
+}) { }
+
+class Plane extends builder.object({
+	shapes: builder.array([Circle, Point])
+}) { }
+
+// ---------------------------------------------------------
+// Change type 1: adding an optional field to an object node
+
+// Suppose the declaration of `Circle` was updated to:
+class Circle extends builder.object({
+	center: Point,
+	radius: builder.number,
+	color: builder.optional(builder.string),
+}) { }
+
+// -------------------------------------------------------------
+// Change type 2: adding to the set of allowed types for a field
+
+// Suppose the application wants to add support for a "square" notion like this:
```

Comment:
This sort of change is backwards compatible with existing _documents_, but not necessarily with existing view schemas. The other 2 classes of changes _are_ compatible with existing view schemas, so it might be worth calling this difference out somewhere?